pkg/controller/core/queue_controller.go

```diff
@@ -47,47 +54,119 @@ func NewQueueReconciler(queues *queue.Manager) *QueueReconciler {
 
 func (r *QueueReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	// No-op. All work is done in the event handler.
+	var queueObj kueue.Queue
+	if err := r.client.Get(ctx, req.NamespacedName, &queueObj); err != nil {
+		// we'll ignore not-found errors, since there is nothing to do.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	log := ctrl.LoggerFrom(ctx).WithValues("queue", klog.KObj(&queueObj))
+	ctx = ctrl.LoggerInto(ctx, log)
+	log.V(2).Info("Reconciling Queue")
+
+	pendingJobs, err := r.queues.Status(&queueObj)
+	if err != nil {
+		r.log.Error(err, "Failed to retrieve queue status")
+		return ctrl.Result{}, err
+	}
+
+	queueObj.Status.PendingWorkloads = int32(pendingJobs)
+	if err := r.client.Status().Update(ctx, &queueObj); err != nil {
+		r.log.Error(err, "Failed to Update queue")
+		return ctrl.Result{}, err
+	}
 	return ctrl.Result{}, nil
 }
 
 func (r *QueueReconciler) Create(e event.CreateEvent) bool {
-	q := e.Object.(*kueue.Queue)
+	q, match := e.Object.(*kueue.Queue)
+	if !match {
+		// No need to interact with the cache for other objects.
+		return true
+	}
 	log := r.log.WithValues("queue", klog.KObj(q))
 	log.V(2).Info("Queue create event")
 	ctx := logr.NewContext(context.Background(), log)
 	if err := r.queues.AddQueue(ctx, q); err != nil {
 		log.Error(err, "Failed to add queue to system")
 	}
-	return false
+	return true
 }
 
 func (r *QueueReconciler) Delete(e event.DeleteEvent) bool {
-	q := e.Object.(*kueue.Queue)
+	q, match := e.Object.(*kueue.Queue)
+	if !match {
+		// No need to interact with the cache for other objects.
+		return true
+	}
 	r.log.V(2).Info("Queue delete event", "queue", klog.KObj(q))
 	r.queues.DeleteQueue(q)
-	return false
+	return true
 }
 
 func (r *QueueReconciler) Update(e event.UpdateEvent) bool {
-	q := e.ObjectNew.(*kueue.Queue)
+	q, match := e.ObjectNew.(*kueue.Queue)
+	if !match {
+		// No need to interact with the cache for other objects.
+		return true
+	}
 	log := r.log.WithValues("queue", klog.KObj(q))
 	log.V(2).Info("Queue update event")
 	if err := r.queues.UpdateQueue(q); err != nil {
 		log.Error(err, "Failed to update queue in system")
 	}
-	return false
+	return true
 }
 
 func (r *QueueReconciler) Generic(e event.GenericEvent) bool {
 	r.log.V(3).Info("Ignore generic event", "obj", klog.KObj(e.Object), "kind", e.Object.GetObjectKind().GroupVersionKind())
-	return false
+	return true
+}
+
+type queueqWorkloadHandler struct{}
+
+func (h *queueqWorkloadHandler) Create(e event.CreateEvent, q workqueue.RateLimitingInterface) {
+	qw := e.Object.(*kueue.QueuedWorkload)
+	// Check that the wl has been assigned first
+	if workloadStatus(qw) != "" {
+		q.Add(requestForQueueStatus(qw))
+	}
+}
+
+func (h *queueqWorkloadHandler) Update(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	oldQW := e.ObjectOld.(*kueue.QueuedWorkload)
+	if workloadStatus(oldQW) != "" {
```

Comment:
If a workload transitioned from pending to assigned or finished, we need to Renconcile in all cases.

If a workload switched queues, we need to update both queues.

I don't think this function is doing that.